operator/src/main/java/com/instaclustr/cassandra/operator/service/ControllerService.java

```diff
@@ -1,428 +1,435 @@
-            Integer currentReplicas = appsApi.readNamespacedStatefulSet(statefulSet.getMetadata().getName(), namespace, null, null, null).getStatus().getReplicas();
-            if (currentReplicas == 0) 
-                break;
 
-            Thread.sleep(50);
+        // TODO: this will be slow for a large number of nodes -- introduce some parallelism
+        final ImmutableListMultimap<CassandraConnection.Status.OperationMode, V1Pod> podCassandraOperationModes = Multimaps.index(pods, p -> {
+            final CassandraConnection cassandraConnection = cassandraConnectionFactory.connectionForPod(p);
+
+            return cassandraConnection.status().operationMode;
+        });
+
+
+        if (EnumSet.of(CassandraConnection.Status.OperationMode.NORMAL, CassandraConnection.Status.OperationMode.DECOMMISSIONED).equals(podCassandraOperationModes.keySet())) {
+            // change statefulset size
+
+            // TODO: handle/assert that the newest pod is the one that was decommissioned
+
+            statefulSet.getSpec().setReplicas(currentReplicas - 1);
+            appsApi.replaceNamespacedStatefulSet(statefulSet.getMetadata().getName(), statefulSet.getMetadata().getNamespace(), statefulSet, null);
+
+            return;
         }
 
-        logger.debug("done with scaling to 0");
+        if (!EnumSet.of(CassandraConnection.Status.OperationMode.NORMAL).equals(podCassandraOperationModes.keySet())) {
+            logger.warn("Skipping StatefulSet reconciliation as some Cassandra nodes are not in the NORMAL state: {}",
+                    Multimaps.transformValues(podCassandraOperationModes, (V1Pod p) -> p.getMetadata().getName())
+            );
 
-        //Catch JsonSyntaxException to avoid crash (https://github.com/kubernetes-client/java/issues/86)
-        try {
-            deleteResource(
-                    () -> appsApi.deleteNamespacedStatefulSet(statefulSet.getMetadata().getName(), namespace, deleteOptions, null, null, null, "Foreground")
-                    );
+            return;
         }
-        catch (JsonSyntaxException e) {
-            if (e.getCause() instanceof IllegalStateException) {
-                IllegalStateException ise = (IllegalStateException) e.getCause();
-                if (ise.getMessage() != null && ise.getMessage().contains("Expected a string but was BEGIN_OBJECT"))
-                    logger.debug("Catching exception because of issue https://github.com/kubernetes-client/java/issues/86");
-                else throw e;
-            }
-            else throw e;
+
+
+        if (desiredReplicas >= currentReplicas) {
+            // scale up or modify
+            appsApi.replaceNamespacedStatefulSet(statefulSet.getMetadata().getName(), statefulSet.getMetadata().getNamespace(), statefulSet, null);
+
+        } else  {
+            // scale down
+
+            // decommission the newest pod
+            cassandraConnectionFactory.connectionForPod(pods.get(0)).decommission();
```

Comment:
iirc decom will fail when you call it on the last node left. Cassandra will throw an exception

Thinking it through a little further, it doesn't make sense to decom when we are trying to get to zero to just delete the cluster... as the last few nodes will have a ton of data (it will also be slow).

We should probably have a special path for going from N -> 0

Reply:
I think you mentioned we don't handle the delete case as the moment anyway (it leaves the PVs around)?

Reply:
Do we care about scale -> 0? Currently the `CassandraDataCenter` CRD has a minvalue for the `replicas` field of 1, meaning there always has to be one node.

True deletion of a DC is handled separatly -- when the DC object is deleted (e.g. `kubectl delete cdc test-datacenter-1`). In this case we nuke everything -- services, default config maps, generated secrets, etc.

In contrast, StatefulSets do allow a `replicas` value of 0, which results in 0 pods being created.

Handling scale -> 0 wouldn't be difficult, I just wonder if it's necissary.

Reply:
If full deletion of the CDC resource is handled seperately, then this is less of an issue. With the min value of 1 in replicas that also works for me.

LGTM.